File: docs/vsto/codesnippet/VisualBasic/Trin_VstcoreHostControlsExcelVB/Sheet2.vb
Public Class Sheet2

    '--------------------------------------------------------------------------
    Dim WithEvents list1 As Microsoft.Office.Tools.Excel.ListObject


    '--------------------------------------------------------------------------
    '<Snippet12>
    Dim id As Integer = 0
    Dim employeeTable As System.Data.DataTable
    '</Snippet12>


    '--------------------------------------------------------------------------
    Private Sub Sheet1_Startup(ByVal sender As Object, ByVal e As System.EventArgs) _
        Handles Me.Startup
        '<Snippet13>
        employeeTable = New System.Data.DataTable("Employees")

        Dim column As System.Data.DataColumn = employeeTable.Columns.Add("Id", GetType(Int32))
        column.AllowDBNull = False

        employeeTable.Columns.Add("FirstName", GetType(String))
        employeeTable.Columns.Add("LastName", GetType(String))
        employeeTable.Columns.Add("Age", GetType(Int32))

        employeeTable.Rows.Add(id, "Nancy", "Anderson", 56)
        employeeTable.Rows.Add(id, "Robert", "Brown", 44)
        id += 1

        list1.SetDataBinding(employeeTable, "", "FirstName", "LastName", "Age")
        '</Snippet13>
    End Sub



    '--------------------------------------------------------------------------
    '<Snippet14>
    Private Sub list1_ErrorAddDataBoundRow(ByVal sender As Object, ByVal e As  _
        Microsoft.Office.Tools.Excel.ErrorAddDataBoundRowEventArgs) _
        Handles list1.ErrorAddDataBoundRow

        Dim row As System.Data.DataRow = (CType(e.Item, System.Data.DataRowView)).Row

        If TypeOf (e.InnerException) Is NoNullAllowedException Then
            row("Id") = id
            id += 1
            e.Retry = True
        End If
    End Sub
    '</Snippet14>
End Class

File: docs/vsto/codesnippet/CSharp/Trin_VstcoreHostControlsExcelCS/Sheet2.cs
using System;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using Microsoft.VisualStudio.Tools.Applications.Runtime;
using Excel=Microsoft.Office.Interop.Excel;
using Office=Microsoft.Office.Core;

namespace Trin_VstcoreHostControlsExcelCS
{
    public partial class Sheet2
    {
        //---------------------------------------------------------------------
        Microsoft.Office.Tools.Excel.ListObject list1 = null;
        

        //---------------------------------------------------------------------
        //<Snippet12>
        private int id = 0;
        private System.Data.DataTable employeeTable;
        //</Snippet12>


        //---------------------------------------------------------------------
         private void Sheet1_Startup(object sender, System.EventArgs e)
        {
            //<Snippet13>
            employeeTable = new System.Data.DataTable("Employees");

            System.Data.DataColumn column = 
                employeeTable.Columns.Add("Id", typeof(int));
            column.AllowDBNull = false;

            employeeTable.Columns.Add("FirstName", typeof(string));
            employeeTable.Columns.Add("LastName", typeof(string));
            employeeTable.Columns.Add("Age", typeof(int));

            employeeTable.Rows.Add(id, "Nancy", "Anderson", "56");
            employeeTable.Rows.Add(id, "Robert", "Brown", "44");
            id++;

            list1.SetDataBinding(employeeTable, "", "FirstName", "LastName", "Age");

            //<Snippet15>
            list1.ErrorAddDataBoundRow += new Microsoft.Office.Tools.Excel.
                ErrorAddDataBoundRowEventHandler(list1_ErrorAddDataBoundRow);
            //</Snippet15>
            //</Snippet13>
        }
        


        //---------------------------------------------------------------------
        //<Snippet14>
        private void list1_ErrorAddDataBoundRow(object sender,
            Microsoft.Office.Tools.Excel.ErrorAddDataBoundRowEventArgs e)
        {
            System.Data.DataRow row = ((System.Data.DataRowView)e.Item).Row;

            if (e.InnerException is NoNullAllowedException)
            {
                row["Id"]= id;
                id++;
                e.Retry = true;
            }
        }
        //</Snippet14>
        //---------------------------------------------------------------------
        private void InternalStartup()
        {
            this.Startup+=new System.EventHandler(Sheet1_Startup);
        }
    }
}
